Improve waiting for messages on Windows

AngelicosPhosphoros opened the pull request:
Previous version used [`SetTimer`] with `GetMessageW` for waiting.
The downside of UI timers like ones created by `SetTimer`,
is that they may be late by up to 15-16 ms.

To fix this behaviour, I added use of high resolution timers created by [`CreateWaitableTimerExW`] with the flag `CREATE_WAITABLE_TIMER_HIGH_RESOLUTION`.
In my previous experience, waiting on such timers have precision of roundly 0.5 ms which is the best available on Windows at the moment.
I use [`MsgWaitForMultipleObjectsEx`] to wait simultaneously for both timer and newly arriving events.

Unfortunately, high resolution timers are available only since Windows 10 1803. However:

1. Win 10 is already getting to the end of support, like all previous versions, so it is OK to rely on APIs introduced in it;
2. I use `dwMilliseconds` parameter of `MsgWaitForMultipleObjectsEx` as a fallback. It should perform not worse compared to waiting for events from `SetTimer`.

I also refactored code to remove event dispatching from function responsible for waiting for events. This provides more clear separations of concern and avoids unnecessary duplication of dispatching logic.

After [review] from @rib, I also moved the waiting itself from `wait_for_messages` method to separate function, so it is clearly seen that `wait_for_messages` do 3 things: notify app that we about to wait, wait, notify that we have new events.

I have tested behaviour using a egui app with Vulkan rendering with `VK_PRESENT_MODE_IMMEDIATE_KHR`, and older version consistently have twice less FPS than requested (e.g. 30 FPS when limit is 60 and 60 FPS when limit is 120) while newer version works more correctly (almost always 60 FPS when limit is 60, and only 5-10 frames missing when FPS is set to 120 or more).

Fixes https://github.com/rust-windowing/winit/issues/1610

[`CreateWaitableTimerExW`]: https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createwaitabletimerexw
[`MsgWaitForMultipleObjectsEx`]: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-msgwaitformultipleobjectsex
[`SetTimer`]: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-settimer
[review]: https://github.com/rust-windowing/winit/pull/3950#discussion_r1800184479

- [x] Tested on all platforms changed
- [x] Added an entry to the `changelog` module if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

AngelicosPhosphoros commented:
#### Current version in crates.io:

https://github.com/user-attachments/assets/bf52ce3e-c05d-4afa-b8c4-a20784b0e98d

#### With my changes:

https://github.com/user-attachments/assets/412e502d-ca59-43b8-a222-1778fa043553

AngelicosPhosphoros commented:
Rebased on top of master branch.

AngelicosPhosphoros commented:
I may be unavailable during a week but if you have questions/comments, I would happily address them in next Saturday.

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -127,6 +134,10 @@ pub(crate) enum ProcResult {
+    // It is a timer used on timed waits.
+    // It is created lazily in case if we have `ControlFlow::WaitUntil`.
+    // Keep it as a field to avoid recreating it on every `ControlFlow::WaitUntil`.
+    high_precision_timer: Option<HANDLE>,
```
Can we use [`OwnedHandle`](https://doc.rust-lang.org/std/os/windows/io/struct.OwnedHandle.html) for this instead to simplify cleanup?

AngelicosPhosphoros replied:
Ok

AngelicosPhosphoros replied:
Honestly, it looks much messier as result because types used in `std` and in `windows-sys` crate are not compatible.
Btw, why use `windows-sys` crate instead of `windows` crate? `windows` crate does all error checking itself and has more ergonomic interface overall.

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -245,17 +257,13 @@ impl EventLoop {
         }
 
         let exit_code = loop {
-            self.wait_and_dispatch_message(None);
-
-            if let Some(code) = self.exit_code() {
-                break code;
-            }
-
             self.dispatch_peeked_messages();
 
             if let Some(code) = self.exit_code() {
                 break code;
             }
+
+            self.wait_for_messages(None);
```
Please clarify why we've removed the `exit_code()` check between waiting for messages and dispatching peeked messages.

AngelicosPhosphoros replied:
As I understand, it was here to handle cases when exit codes were set in callbacks to messages that called in `wait_and_dispatch_message`. I don't do any dispatch in new `self.wait_for_messages` so there is no new exit code that can appear before `dispatch_peeked_messages`. Also, before the loop, exit code is reset at https://github.com/rust-windowing/winit/blob/c858ac7d157535250a2ea35deb70b9e687bd4a86/src/platform_impl/windows/event_loop.rs#L231 so it cannot be set before first iteration of loop either.

rib replied:
It seems notable that the ordering of `dispatch_peeked_` followed by `wait_for_messages` (instead of waiting then dispatching) here isn't consistent with the `pump_app_events` implementation, which seems like it could affect the order of dispatching some initial events for an app.

I believe `wait_for_messages` dispatches the `AboutToWait` event (`runner.prepare_wait()`) to notify the app that it's about to block and wait for input so I wonder if it might still make sense to check for a possible exit code from the app after `wait_for_messages()`

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -343,40 +350,65 @@ impl EventLoop {
-                }
-                // A return value of 0 implies `WM_QUIT`
-                if get_status == 0 {
-                    PumpStatus::Exit(0)
+                let handle: HANDLE = CreateWaitableTimerExW(
+                    ptr::null(),
+                    ptr::null(),
+                    CREATE_WAITABLE_TIMER_HIGH_RESOLUTION,
+                    TIMER_ALL_ACCESS,
+                );
+                // CREATE_WAITABLE_TIMER_HIGH_RESOLUTION is supported only after
+                // Win10 1803 but it is already default option for rustc
+                // (std uses it to implement `std::thread::sleep`).
+                if handle == 0 {
+                    None
                 } else {
-                    PumpStatus::Continue
+                    Some(handle)
                 }
             }
         }
```
I would rather make these freestanding functions than having them inlined into `wait_for_messages`.

AngelicosPhosphoros replied:
OK, would fix tomorrow. I just kept the structure that was used before.

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -393,57 +425,84 @@ impl EventLoop {
-            },
-        };
-        let timeout = min_timeout(control_flow_timeout, timeout);
+        {
+            let control_flow_timeout = match runner.control_flow() {
+                ControlFlow::Wait => None,
+                ControlFlow::Poll => Some(Duration::ZERO),
+                ControlFlow::WaitUntil(wait_deadline) => {
+                    let start = Instant::now();
+                    Some(wait_deadline.saturating_duration_since(start))
+                },
+            };
+            timeout = min_timeout(timeout, control_flow_timeout);
+        }
```
```suggestion
        let timeout = {
            let control_flow_timeout = match runner.control_flow() {
                ControlFlow::Wait => None,
                ControlFlow::Poll => Some(Duration::ZERO),
                ControlFlow::WaitUntil(wait_deadline) => {
                    let start = Instant::now();
                    Some(wait_deadline.saturating_duration_since(start))
                },
            };
            min_timeout(timeout, control_flow_timeout)
        };
```

Slightly cleaner this way

AngelicosPhosphoros replied:
Should I declare new binding `let timeout` or just update existing variable `timeout`? I would update code tomorrow.

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -393,57 +425,84 @@ impl EventLoop {
-        // API) and there's no API to construct or initialize a `MSG`. This
-        // is the simplest way avoid uninitialized memory in Rust
-        let mut msg = unsafe { mem::zeroed() };
-        let msg_status = wait_for_msg(&mut msg, timeout);
```
Why did we stop using `wait_for_msg`?

AngelicosPhosphoros replied:
Because I rewrote code to use `MsgWaitForMultipleObjectsEx` instead of `GetMessageW`. `MsgWaitForMultipleObjectsEx` only wait until a new message is available but doesn't take it and doesn't remove it from queue, while `GetMessageW` does wait and take it.

In new version, `wait_for_messages` only waits until message is available, then `dispatch_peeked_messages` takes them from queue and dispatches, until the queue is empty.

So, as I understand, old version of code needed to manage both waiting and dispatching messages so it was split it into subfunctions. New version of code only does waiting while all dispatching done by `dispatch_peeked_messages` so there is no need to create a subfunction which just repeats the name of outer function `wait_for_messages`.

Also, after I move inner functions out of `wait_for_messages`, as you requested in https://github.com/rust-windowing/winit/pull/3950#discussion_r1798559529 , `wait_for_messages` would be tiny and splitting it further would be pointless.

notgull replied:
I believe there are some messages that `MsgWaitForMultipleObjectsEx` does not receive, which is why we used the timer system and `GetMessageW` here. 

cc @rib you were the one to set up this system

rib replied:
I don't really recall there being a particular issue with `MsgWaitForMultipleObjectsEx` and at the time I suppose it just seemed simpler to use `GetMessage` while we were only waiting for messages and not waiting on any object handles.

The main concern I had when last iterating on this code was to do with the indirection of a dedicated waiter thread and there were some funky details to do with redraw handling I don't really recall.

Ref: fae4cbd2aa67315b29f139142bb54a5ff358b95c and https://github.com/rust-windowing/winit/pull/2767

This change to use hrtimers seems good to me, and can see that it makes sense to use `MsgWaitForMultipleObjectsEx` to integrate the timer handle.

It seems to make sense that the waiting is now decoupled, since `MsgWaitForMultipleObjectsEx` doesn't take any data when it has something to wake up for - unlike `GetMessageW` which would wait _and_ take.

notgull reviewed (changes requested):
Initial review

filnet commented:
The timer resolution issue was fixed in Feb 2021 : https://github.com/rust-windowing/winit/commit/cfbe8462ccd7d442f1dd7aa4807adf8b3e75e8e3
It used `timeapi::timeBeginPeriod` and `timeapi::timeEndPeriod` as described here : https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/

It was later removed in July 2023: https://github.com/rust-windowing/winit/commit/420840278b5d708e34b10194aba02ce37e9ac7bf#diff-bb05c5e1236bdaa7f268bcbbbb8cc9f378a48eb3ed480884f0c6471a9cc933edL467

It is not clear why it was removed, the commit does not mention that change.
Just sayin...

AngelicosPhosphoros commented:
@filnet Well, `timeapi::timeBeginPeriod` and `timeapi::timeEndPeriod` should generally NOT be used because it causes big increase of energy consumption. It makes ALL system thread scheduler run more frequently which affects all programs and cause more frequent thread rescheduling. And it is still less precise compared to CREATE_TIMER_HIGH_PRECISION.

As for the issue being fixed, well, I see that it is still open. And as can be seen in attached video, currently WaitUntil just don't work correctly (e.g. getting 30 FPS when I ask for 60).

AngelicosPhosphoros commented:
@notgull I addressed all comments except about missing messages by `MsgWaitForMultipleObjectsEx`. I would look which ones can be missed myself and wait for @rib .

AngelicosPhosphoros commented:
I just noticed that I used word "precision" instead of "resolution" everywhere. Fixed it now.

rib commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -343,107 +350,115 @@ impl EventLoop {
 
+        // Before we potentially exit, make sure to consistently emit an event for the wake up
+        runner.wakeup();
+
```
Maybe this should be coupled with the implementation of `wait_for_messages`. 

Conceptually the intention was to have a sequence like:
1. notify the app that it's about to block/wait
2. _use OS API to block_
3. notify the app of a wake up
... then dispatch/handle all events that triggered the wake up

So long as `dispatch_peeked_messages` only gets called once immediately after `wait_for_messages` returns then notifying the wake up here is OK, but it feels to me like it would be more correct to notify about the wakeup immediately after `MsgWaitForMultipleObjectsEx` returns and avoid any risk of dispatching multiple wake ups if someone makes a change later that ends up calling `dispatch_peeked_messages` multiple times per wait.

filnet commented:
> @filnet Well, `timeapi::timeBeginPeriod` and `timeapi::timeEndPeriod` should generally NOT be used because it causes big increase of energy consumption. It makes ALL system thread scheduler run more frequently which affects all programs and cause more frequent thread rescheduling. Also, I think, it also reduces time slices that each thread receive when scheduled to execution. And it is still less precise compared to CREATE_TIMER_HIGH_RESOLUTION.
> 
> As for the issue being fixed, well, I see that it is still open. And as can be seen in attached video, currently WaitUntil just don't work correctly (e.g. getting 30 FPS when I ask for 60).

I just wanted to mention the previous fix for completeness sake... If there is a better solution then we should use it.
The issue is still open because the previous fix was removed. Which issue btw ?

AngelicosPhosphoros commented:
@filnet Sorry if my previous message come out as a too forceful, just I wanted to make downsides of changing timer frequency clear. If you want to read what the problems with changing frequency timer, you may read this article: https://randomascii.wordpress.com/2013/07/08/windows-timer-resolution-megawatts-wasted/

AngelicosPhosphoros commented:
I rebased on top of master branch and also made changes requested by at https://github.com/rust-windowing/winit/pull/3950#discussion_r1800184479

AngelicosPhosphoros commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -312,8 +325,11 @@ impl EventLoop {
-
+        if self.exit_code().is_none() {
+            self.wait_for_messages(timeout);
+        }
```
I have some concern here. Right now, it looks like this sequence happens here:
1. Call to `runner.wakeup()`.
2. Call to `wait_for_messages`
3. Inside `wait_for_messages` immeadiately calls `runner.prepare_wait()`.
4. Waits for events.
5. Call `runner.wakeup()`
6. `wait_for_messages` finishes
7. dispatch events
8. Call `runner.prepare_wait()`

My concern that there is no check if there are any new events between first `wakeup` and first `prepare_wait`. Maybe we should do sequence of `wakeup`. dispatch, `wait_for_messages`, dispatch, `prepare_wait`?

@rib I tag you because you previously asked to change calls to `wakeup` and `prepare_wait`.

rib replied:
Hi, sorry that I don't have much time to follow the details here.

If we say that `prepare_wait` and `wakeup` are conceptually supposed to surround the "OS wait point" then it looks like we currently have two competing definitions of what that OS wait point should be...

It looks `pump_app_events` is intending to define the relinquish of control to the application loop (via `PumpStatus::Continue`) as the point where the Winit event loop will wait.

This arguably makes sense because `pump_app_events` is supposed to be non-blocking and any application using `pump_app_events` in their own loop should do something to throttle that loop before repeated called to `pump_app_events` (so applications using `pump_app_events` are arguably responsible for the wait, and they would also be expected to call `pump_app_events` after they wake up).

Sorry if I confused matters by overlooking these details when first commenting about changes here.

It looks like this needs to pick one definition.

Either:
1. Define the wait point by the act of relinquishing control to the app via `::Continue` and have some argument for `wait_for_messages()` that tells that function to skip over `prepare_wait` and `wakeup` when used from here.
2. Define the wait point as the call to `wait_for_messages_impl` and instead remove the call to `.wakeup()` at the start of `pump_app_events` and remove the call to `prepare_wait` just before we return with a `::Continue` status.

I guess option 1 would be most consistent with the current semantics.

AngelicosPhosphoros replied:
I think, maybe we should fix it in another PR? This PR doesn't change behaviour at this point and it is focused on fixing waiting precision.

AngelicosPhosphoros commented:
CI failure seems to be irrelevant because it is about another platform (Redox).

AngelicosPhosphoros commented:
@notgull Just reminding that I still wait for a code review.

notgull commented:
I will review this today or tomorrow.

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -668,10 +618,136 @@ fn dur2timeout(dur: Duration) -> u32 {
+    }
+}
+
+/// Implementation detail of [EventLoop::wait_for_messages].
+/// Does actual system-level waiting and doesn't process any messages itself,
```
```suggestion
/// Implementation detail of [EventLoop::wait_for_messages].
/// 
/// Does actual system-level waiting and doesn't process any messages itself,
```

AngelicosPhosphoros replied:
Done.

notgull commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -668,10 +618,136 @@ fn dur2timeout(dur: Duration) -> u32 {
+    }
+}
+
+/// This function should not return error if parameters valid
+/// but there is no guarantee about that at MSDN docs
+/// so we return result of GetLastError if fail.
+/// ## Safety
+/// timer must be a valid timer handle created by create_high_resolution_timer.
+/// timeout divided by 100 nanoseconds must be more than 0 and less than i64::MAX.
```
```suggestion
/// This function should not return error if parameters valid
/// but there is no guarantee about that at MSDN docs
/// so we return result of GetLastError if fail.
/// 
/// ## Safety
/// 
/// timer must be a valid timer handle created by create_high_resolution_timer.
/// timeout divided by 100 nanoseconds must be more than 0 and less than i64::MAX.
```

AngelicosPhosphoros replied:
Done.

notgull reviewed (changes requested):
LGTM aside from some minor nits

AngelicosPhosphoros commented:
@notgull I rebased on top of master branch and applied proposed changes.
